Reduce PWA's dependencies

vmarta opened the pull request:
As part of speeding up the project-generation time, we wanted to audit the Retail React App's dependencies because installing them proves to be the bottleneck. 

Ticket: https://gus.lightning.force.com/a07EE00000cJvxkYAC

This PR removes dependencies that fall under these categories:
- Which dependencies are actually no longer used?
- Which dependencies are used but probably not crucial?
  - The biggest one that falls in this category is Cypress. And Kevin has taken it out recently #294.
  - There are 3 different kinds of `fetch` libraries we used over time. So I replaced `node-fetch` with `cross-fetch`. There’s also `whatwg-fetch` but it looks to be required by our webpack config, so I leave it in just to be safe.

The general process I used to determine which dependencies to remove is:
1. First, run `depcheck` tool and look at the result
2. For each dependency in the result:
    1. Search for it in the entire codebase
    2. Google the purpose of the dependency
    3. Check the git history in the [old repo](https://github.com/mobify/mobify-platform-sdks/tree/salesforce-develop/packages/pwa) for some context
    4. If I decided to remove this dependency, I’d also double check our CI test result to confirm everything’s ok



# Description

<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relavant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [x] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# How to Test-Drive This PR

- Verify that the CI tests are passing
- Smoke test the pwa 

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [x] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [x] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

vmarta committed:
Remove unused pwa dependencies

vmarta committed:
PWA: replace node-fetch with cross-fetch

So we no longer install both libraries.

vmarta committed:
Update CHANGELOG.md

bendvc commented on packages/pwa/package.json:
```diff
@@ -113,16 +109,12 @@
-        "minimist": "1.2.5",
         "msw": "^0.28.1",
         "njwt": "^1.0.0",
         "nock": "^13.1.1",
```
I don't see `nock` being used in the PWA, but I see that it's defined as a common dependency in said script. Is there something we can do here.. potentially not make it a common dep? I'm not 100% familiar with that script so we'd want to be cautious if we make changes there.

vmarta replied:
Yup, I'm not familiar with the script either. I see that `nock` is only used within the sdk package and nowhere else. So it seems like we can remove it for all packages except the SDK. 

But... again, I'm not familiar with the context of the script. So I've created a [ticket](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000i1IPtYAM/view) for us to look at this in the future.

bendvc commented:
As I mentioned in [slack](https://mobify.slack.com/archives/C0DKK1FJS/p1642452585047500?thread_ts=1642450241.046000&cid=C0DKK1FJS) these types of problems aren't fixable by doing what we are doing. We almost need to include guard rails to ensure that we don't get out of had. This might manifest itself in a dependency checker that ensures the project doesn't 1. have to many and 2. doesn't have any that aren't used. 

Are we working on a PR that will doing something like that?

vmarta committed:
Merge branch 'develop' into audit-pwa-dependencies

vmarta committed:
Update package-lock.json

vmarta committed:
Update max-number-of-packages we're allowing for

vmarta committed:
Fix a bug with setting environment variable

After setting env variable, we need to make sure to source $BASH_ENV immediately after.

vmarta commented on .circleci/config.yml:
```diff
@@ -294,15 +294,16 @@ jobs:
           environment:
           command: |
-            MAX_PACKAGES="3050"
+            MAX_PACKAGES="2250"
             total=$(./scripts/count-dependencies.js /tmp/project)
```
Credit to Oli for mentioning this script in a Slack thread recently.

vmarta commented on .circleci/config.yml:
```diff
@@ -294,15 +294,16 @@ jobs:
+            MAX_PACKAGES="2250"
             total=$(./scripts/count-dependencies.js /tmp/project)
             echo "export TOTAL_PACKAGES=${total}" >> $BASH_ENV;
+            source $BASH_ENV
```
Fix for a regression bug where we forgot to apply/source the new environment variable `TOTAL_PACKAGES`.

vmarta replied:
@adamraya since you have access to Datadog, can you check what happened to the data lately? For `generated_project_total_packages`, whether any value is recorded for it or not. Because it looks like we've been sending blanks.

FYI this is where we sent the data from: https://github.com/SalesforceCommerceCloud/pwa-kit/blob/7bede2aae4787f62061b52b0cc423a78f5539cd3/.circleci/config.yml#L341

bendvc replied:
Good catch with the missing `source` call.

adamraya replied:
It looks like Datadog was still receiving data. But it could be that the data is coming from generated projects not using the latest changes, so it's a good thing to fix.

vmarta commented:
> As I mentioned in [slack](https://mobify.slack.com/archives/C0DKK1FJS/p1642452585047500?thread_ts=1642450241.046000&cid=C0DKK1FJS) these types of problems aren't fixable by doing what we are doing. We almost need to include guard rails to ensure that we don't get out of had. This might manifest itself in a dependency checker that ensures the project doesn't 1. have to many and 2. doesn't have any that aren't used.
> 
> Are we working on a PR that will doing something like that?

To address point 1, there are actually 2 existing methods for it:
- We have a script that checks for the number of packages. In this PR, I've just updated the max number that we allow for it.
- There's a new [ticket](https://gus.lightning.force.com/a07EE00000hv97ZYAQ) to bring back the bundle-size test. It looks like we intended to remove it for Windows testing only, but ended up removing for all.

As for addressing point 2, the tool `depcheck` is great for it but unfortunately, depcheck has quite many false positives in my experience. So I think it's best if it's done manually at regular interval. Or if we want to automate it, we may need an allowed list to filter out depcheck's result.

vmarta commented:
@bendvc please review this PR again when you have time, thanks.
